### ISSUE TITLE
Retrofit 3.x Things

### ISSUE THREAD

**Issue by JakeWharton** [square/retrofit#2180]
Currently no plans for a Retrofit 3.x. Just tracking changes in a single issue instead of many.

 * Mock behavior should invoke the implementation method for each `Call.execute()` / `Call.enqueue()`. Currently it only invokes it once and reuses the result for every cloned `Call`.

**Comment by TobiasReich** [comment]
Will there be a simple way in order to retry a call n times?
This would come very handy when there is a call that should be retried after an error.
I was hoping for something like: request.retry(5);

**Comment by naturalwarren** [comment]
To do this with Observables we use a custom transformer that gets tacked onto request invocations that takes in the number of retries.

**Comment by Jawnnypoo** [comment]
Curious if for a 3.x it would be appropriate to have `HttpException` be a part of the `retrofit` module, instead of creating it separately in each adapter that is created. Is there a reason for this currently?

**Comment by JakeWharton** [comment]
There's no use for it in the main module at present which is why each
adapter defines its own. This change wouldn't need to wait for 3.x as it
can be added in a backwards compatible-way.

**Comment by Jawnnypoo** [comment]
Gotcha. So do you think that each should continue to have its own copy? The benefits I could see with unifying mostly come from just reducing the duplication. Is this something you would be interested in seeing in a PR, or do you prefer keeping them separate?

**Comment by JakeWharton** [comment]
I'd be okay with a PR. Basically copy paste it into the main module and
then change all the existing ones to extend. I'm not happy it has to be
non-final in the main module, but it's only an exception so not a big deal.

**Comment by JakeWharton** [comment]
I created #2194 to track that. Feel free to send a PR over the weekend. Otherwise I'll grab it on Monday. I need to release really soon but I'll wait for this.

**Comment by JakeWharton** [comment]
@TobiasReich no plans for retry. You can already call `clone()` on a `Call` to retry and so implementing that at the application level seems more appropriate. You can write a `CallAdapter` to return a custom type that provides a nice API for retry. Because of the behavior in sync vs. async mode and what executor the delay happens on being unclear I don't think I want it in the library.

**Comment by soodakshay** [comment]
Hey @JakeWharton  I am a big fan of yours. Can I get some code sample so that I can learn a good coding style? I am an Android Developer and I want to improve my coding styles and learn about the quality code.
Hope you will help me
Thanks :)

**Comment by ylfzq** [comment]
You can add code below to your okhttp interceptor to archieve retry mechanism.

```java
new OkHttpClient.Builder()
  .addInterceptor(new RetryInterceptor(3)) // must add at first
  // ...
  .build();
```

The RetryInterceptor is
```java
public class RetryInterceptor implements Interceptor {
  private final int mRetryTimes;

  public RetryInterceptor(int retryTimes) {
    mRetryTimes = retryTimes;
  }

  @Override
  public Response intercept(Chain chain) throws IOException {
    int retryTimesLeft = mRetryTimes;
    for (; retryTimesLeft > 0; --retryTimesLeft) {
      try {
        return chain.proceed(chain.request());
      } catch (IOException ignore) {
      }
    }
    throw new IOException("still fail after retry " + mRetryTimes + " times");
  }
}
```

The code is not tested yet. But I think it's quite simple to understand.

**Comment by Skullper** [comment]
**Kotlin**
Will retrofit support async/await in future?

**Comment by JakeWharton** [comment]
It already does https://github.com/gildor/kotlin-coroutines-retrofit

**Comment by msangel** [comment]
Return easy error api from Retrofit1.

**Comment by jaredsburrows** [comment]
@JakeWharton Can we remove `HttpException` in 3.x? Since it has been unified across the modules here: https://github.com/square/retrofit/commit/4546d5d88a331db779e72fbf453bc5abd69edd2d.

**Comment by JakeWharton** [comment]
Everything deprecated gets removed in major version bumps.

**Comment by jaredsburrows** [comment]
@JakeWharton Great.

**Comment by msangel** [comment]
Thanks. That's what I was asked for.

**Comment by JakeWharton** [comment]
It wasn't what you were asking for. The error handling of 1.x was terribly
designed (by me) and will not be returning.

**Comment by msangel** [comment]
Most of people get http error with adequate, well-formatted message. Yep, not all. But still, most cases the response can be converted well with built-in or custom converters, like those for 2xx responces. And this works perfect in version1. People need this. On stackowerflow I saw workarounds like creating new retrofit instance(or getting instance as singletons) just for accessing converters. Or others solutions(http://bytes.babbel.com/en/articles/2016-03-16-retrofit2-rxjava-error-handling.html)

**Comment by JakeWharton** [comment]
Your assumptions are incorrect. The samples show how to correctly handle
the case where you want to deserialize error responses using a converter.

**Comment by jaredsburrows** [comment]
@msangel Here is the sample: https://github.com/square/retrofit/blob/master/samples/src/main/java/com/example/retrofit/ErrorHandlingAdapter.java.

**Comment by JakeWharton** [comment]
That one doesn't really cover it. I guess there isn't a sample for it. It's a combination of https://github.com/square/retrofit/blob/e6a7cd01657670807bed24f6f4ed56eb59c9c9ab/samples/src/main/java/com/example/retrofit/DeserializeErrorBody.java and https://speakerdeck.com/jakewharton/making-retrofit-work-for-you-droidcon-uk-2016?slide=214

**Comment by msangel** [comment]
Yep. The first sample is good. It is worst then retrofit1 solution just in that moment that error know nothing about it's context. Adding to the error (transient) link to the retrofit instance it was created with will solve the gap. And maybe adding method like `getErrorBodyAs(SomeType.class)` to the HttpError will fulfill it completely.

**Comment by JakeWharton** [comment]
That is not possible because of serialization requirements.

**Comment by vidhithakrar** [comment]
Any plan to support custom annotation on an interface in 3.x?

**Comment by JakeWharton** [comment]
I don't know what that means

**Comment by zergtmn** [comment]
Perhaps he meant something like:
```java
@BaseUrl("https://host1.mydomain.com")
public interface MyService1 {
    @GET("get_something")
    Call<Something> getSomething();
    // ... Many other methods that use the same baseUrl
}
```
Where @BaseUrl is a user-defined annotation. Then CallAdapter.Factory should receive not just method annotations but also the annotations declared on the interface.
I can imagine why this could be useful. Instead of creating a customized Retrofit client before calling `retrofit.create(X.class)` one could create a custom CallAdapter.Factory that would parse those annotations and provide an accordingly customized Retrofit instance to a CallAdapter.

**Comment by vidhithakrar** [comment]
@zergtmn Yeah, something like that. One more thing to add is we get Method level annotation in the CallAdapterFactory but not the Method Parameter level.

public interface FooServices {
    @GET("/path")
    Call<Foo> foo(@Custom String custom);
}

We do not get

**Comment by JakeWharton** [comment]
I see no use to doing that. What would you use it for?

**Comment by heinrichreimer** [comment]
One major point on my whish-list for Retrofit 3 is better type converters.
Moshi [does this excellently](https://github.com/square/moshi#custom-type-adapters) and Google's Room is by far the [most intuitive solution](https://developer.android.com/topic/libraries/architecture/room.html#type-converters).
I know that Retrofit is a bit different in that it features different adapters for the request body and for strings, but it feels like the converter API needs some improvement.
Think of use cases where you want to have one type converter convert from `ResponseBody` to `Foo` and another that can then convert from `Foo` to `Bar`. That way you could e.g. parse.
I've personally worked on a [conversion API](https://github.com/heinrichreimer/android-inquiry/tree/master/library/src/main/java/com/heinrichreimer/inquiry/convert) some time ago and would be happy to help in the development of Retrofit 3.

**Comment by msangel** [comment]
Custom annotation example: interface with the filename as a parameter. We are registering a custom handler for methods with such annotations. Then when the method gets called - the execution will be delivered to the handler. So I will read the file, will process it(compress in my case) and will send in the custom manner. 
And I do not need to write any custom wrapper around the interface just for inject some specific functionality to the generated class. 
Another case: special custom AUTH, when the credentials are just additional request parameters(quite popular case), so we will write a custom handler that will set the credentials. So setting custom annotation and writing handler for it will solve the problem. 
Another case: we can write one custom annotation for queries to one host, and the second - for queries to the second host. 
I believe this annotation should be able to change not only request but the response as well. Actually, you have already builtin logic for convert the response via converters, its good, but still few things there are missing (current `retrofit` object);

**Comment by D330** [comment]
Tag in request

**Comment by Bmooij** [comment]
Honoring the dispatcher.setMaxRequests for observables.
See [#1936](https://github.com/square/retrofit/issues/1936)

**Comment by JakeWharton** [comment]
Already works. Use createAsync().

**Comment by derekm** [comment]
It would be nice if Retrofit entirely adopted JAX-RS for its interface definitions in 3.0. Perhaps a compat jar could be made that keeps the proprietary annotations working as Retrofit moves on to use the standard annotations.

MicroProfile Rest Client 1.0 has entered released candidate status, and it is another JAX-RS-based proxy client. The other JAX-RS-based proxy client I've used extensively is Jersey Proxy Client (`WebResourceFactory`).

I personally much prefer the JAX-RS approach over Retrofit's proprietary annotations. I like to reuse my interface definitions on both the client and the server sides.

**Comment by heinrichreimer** [comment]
I like the idea of implementing the JSR-RS annotations!
If you use exactly the same interface definition and standard on both the server and client side there's not much room for bugs and errors anymore.
And it shouldn't be too hard either as these annotations work very similar to those from Retrofit.

**Comment by nealsanche** [comment]
Allow DELETE with @Body without @HTTP hoop jumping? :)

**Comment by JakeWharton** [comment]
That's not a breaking change and therefore needn't be discussed here.

**Comment by johnrengelman** [comment]
@JakeWharton I've also recently found a use case for custom annotations in the retrofit module for Ratpack. The goal is to allow specifying customer parameters for each individual request type, such as connect/read timeout, max body size, etc.

Additionally, I've found myself limited in trying to implement this, because I can only get the annotations in the `CallAdapterFactory` but the actual binding to the Ratpack `HttpClient` happens in the `CallFactory` and these 2 factories use different types. `CallAdapaterFactory` deals with `retrofit2.Call`, where as `CallFactory` deals with `okhttp3.Call`.
This is an annoying inconsistency in the API.

I know Retrofit has been designed around OkHttp for a while, but with a few changes it could be decoupled entirely from the backing http client implementation.

**Comment by JakeWharton** [comment]
It actually was completely decoupled in 1.x and we determined there was no
value in the additional abstraction and allocation cost. What you're asking
for could be achieved today without a breaking API change by introducing
our own CallFactory for producing OkHttp Call instances.

**Comment by johnrengelman** [comment]
@JakeWharton that would be great for my use case. Would love to see that in a 2.x release :)

**Comment by Edward608** [comment]
Could there be a `@Throw(CustomException)` for individual endpoint?

In my case I want to have a unique exception for each api indicating their failure instead of general `HttpException`, but for now I can only use the work around by `apiCall.onErrorResumeNext { emit(CustomException) }`

Its just my personal opinion, would love to see your feedback about this.

**Comment by JakeWharton** [comment]
You can write a custom CallAdapter for that.

**Comment by Edward608** [comment]
@JakeWharton Could you give me a little bit more hint about this? I see the `Annotation[]` is being passed to the CallAdapterFactory, so this should be the only place to read those annotations?

**Comment by maher640** [comment]
how about executing list of requests in one go and waiting for the data from all requests.

**Comment by KangoV** [comment]
Regarding JAX-RS annotations, Retrofit could handle them how Guice handles javax.inject. It could look for them on the classpath/modulepath and if not there, will not even look for them. We use Ratpack and Retrofit together for our services (no JEE in sight) and would be a shame to have to include a JAX-RS dependency.

**Comment by derekm** [comment]
JAXRS devs are starting discussions on JAXRS 2.2. The 2.2 spec will likely include proxy client details that will set about obsoleting Retrofit.

See, for example:
https://dev.eclipse.org/mhonarc/lists/jaxrs-dev/msg00261.html
... and ...
https://github.com/eclipse-ee4j/jaxrs-api/issues/56
... and ...
https://github.com/eclipse-ee4j/jaxrs-api/issues/598

Retrofit (& Moshi) would do well to align themselves with these emerging standards rather than remaining in an incompatible, proprietary niche API space.

**Comment by technoir3** [comment]
@derekm if Retrofit adopted JAX-RS how it would work on Android? There is no JAX-RS on Android.

**Comment by derekm** [comment]
Retrofit wouldn't need to implement the server-side of JAX-RS. Retrofit 3 could easily be a client-side only implementation of JAX-RS (e.g., https://docs.oracle.com/javaee/7/api/javax/ws/rs/client/package-summary.html + whatever is come up with JAX-RS 2.2's Retrofit/MP-Rest-Client/WebResourceFactory-alike proxy client), with optional jars for backcompat with previous Retrofit releases.

With MP Rest Client 1.0 & JAX-RS 2.2 standardizing proxy clients, what you will start to see is server-side projects offering up JAX-RS/JSON-P/JSON-B/JAXB interface jars for client-side consumption. Retrofit should make itself compatible with these emerging no-code & code-light interfaces & clients.

**Comment by hpanahiird** [comment]
please add a way to retry a failed callback with different base url.

**Comment by BugsBunnyBR** [comment]
Can we be allowed to change the Coroutines dispatcher? It would help a lot with idling resources in UI Tests.

**Comment by JakeWharton** [comment]
No dispatcher is used in our coroutine integration.

**Comment by BugsBunnyBR** [comment]
> No dispatcher is used in our coroutine integration.

Ok, Can we have a way to set the Dispatcher that we want to use when making a request/parsing? So we could centralize this configuration in one place and not at every request and swap the Dispatcher for one that Espresso waits to be idle.

**Comment by JakeWharton** [comment]
The threading is controlled by OkHttp. Retrofit does not have any control over it.

**Comment by BugsBunnyBR** [comment]
So, it means that the parsing is done in the calling thread and not in a background thread? Would this be something that needs to be asked in the parsing libs repository or could it be supported by retrofit 3.0?

**Comment by JakeWharton** [comment]
It is done on the background thread.
